Add wear to contact action

recuero opened the pull request:
## Reason
 - Make wear consideration in contact user-friendly.
 - Avoid looping over (acting on) secondaries with no corresponding mortar segments 

## Design
No changes.

## Impact
Ease of use.


Refs. #19671

recuero committed:
wear in contact action

recuero committed:
Add example using action (#0000)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1064842/) on c4c2b56 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21199/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1064851/) on c4c2b56 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f14a86</th>
      <th colspan=4>#21199 c4c2b5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.56%</td>
      <td>82.56%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75528</td>
      <td>75530</td>
      <td>+2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15955</td>
      <td>15953</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21199/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21199/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>f14a86</th>
      <th colspan=4>#21199 c4c2b5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.00%</td>
      <td>88.96%</td>
      <td>-0.04%</td>
      <td>60.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2914</td>
      <td>2917</td>
      <td>+3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>360</td>
      <td>362</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21199/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21199/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21199/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 60.00% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Add test with wear in contact action (#19671).

Contact pressure and wear depth results match those of an input file building constraint
objects with the aid of the action.

lindsayad commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -78,17 +78,30 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
+      amg().secondariesToMortarSegments(*_current_node);
   std::array<MortarNodalAuxKernelTempl<ComputeValueType> *, 1> consumers = {{this}};
 
+  std::vector<AutomaticMortarGeneration::MortarFilterIter> iterators;
+  for (size_t i = 0; i < its.size(); i++)
+  {
+    mooseAssert(its[i]->first->active(),
+                "We loop over active elements when building the mortar segment mesh, so we golly "
+                "well hope this is active.");
+
+    if (!its[i]->second.empty())
+      // This is local and the mortar segment set isn't empty, so include
+      iterators.push_back(its[i]);
+  }
```
I don't understand this. Do we have cases where there are no mortar segments associated with a secondary? That seems like a problem in the mortar mesh generation or I guess in the generation of this data.

recuero replied:
As the code stands right now, we have secondary elements with no corresponding mortar segment quadrature points (I assume just corresponding mortar segments). Then we make calls in `MortarUtils.h` like this:

``` if (assembly.needDual() )
      assembly.reinitDual(secondary_face_elem, secondary_xi_pts, JxW);
```
      
which require having some mortar segments associated with secondary elements, otherwise we get exceptions from libMesh. 

> Do we have cases where there are no mortar segments associated with a secondary?

I guess in a general case we can have cases where there are no mortar segment associated with a secondary element (say a circle in contact with a small plane, where the circle is secondary surface).

recuero replied:
There is probably a better way to deal with this.

lindsayad replied:
> I guess in a general case we can have cases where there are no mortar segment associated with a secondary element (say a circle in contact with a small plane, where the circle is secondary surface).

In the early days we would have had mortar segments from these, they just wouldn't have any bisections (loose use of bisection here) by inverse projection of primary nodes onto the secondary face element, e.g. the mortar segment elements would have corresponded exactly to the secondary face elements. However, @jbadger95 may have removed these IIRC, e.g. we only have mortar segment elements for cases in which the secondary projects onto the primary. Perhaps this is why `second` can be empty now?

jbadger95 replied:
As far as I remember this is correct @lindsayad. There used to be a `has_primary` flag in mortar segments but because there was a lot of confusion around edge dropping we changed it so that only mortar segments with a primary element are added to the mortar segment mesh.

lindsayad replied:
@recuero would you want to check and see whether you can eliminate the presence of "empty" secondaries from the data returned by `secondariesToMortarSegments` so that we don't have to do manual manipulation in user code (like in `MortarNodalAuxKernel`)? If you don't have time, I can look into it

recuero replied:
Yeah, that sounds like the way to go. I can look into that. Give me a bit of time.

lindsayad committed:
Only evaluate mortar nodal aux kernel if var is nodal defined

recuero committed:
Eliminate presence of empty secondary elements.

There will be nodes that do not project onto the primary surface. This will
cause that maps in loopOverMortarSegments that do not have elements in the mortar segment
mesh. As such, no integration points are identified, which will cause issues when reiniting
entities such as dual bases. Here we eliminate empty secondaries in AutomaticMortarGeneration::secondariesToMortarSegments.

recuero committed:
Avoid mortar auxiliary variable NaN if it doesn't project.

lindsayad committed:
Remove secondary elements without msms from map

lindsayad committed:
Remove unused secondaries from node-to-elem maps

lindsayad commented:
@recuero feel free to review my changes

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1095169/) on 1cf3838 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)

recuero commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -2230,24 +2241,21 @@ AutomaticMortarGeneration::secondariesToMortarSegments(const Node & node) const
-      // We may have removed this element key from this map
-      continue;
-
+      mooseError("end of mortar segment map in associatedMortarSegments");
```
Each secondary element in the maps must now always have a corresponding mortar segment by mesh construction (my understanding).

recuero replied:
We are removing the secondaries from the nodal map if the msm is empty: Is that general enough to enforce this error?

lindsayad replied:
Yes it should be. We should not have any secondaries without msms either in the node-to-secondary map or the secondary-to-msm map.

recuero replied:
We may be missing something. I created a typical bouncing block example with a secondary sideset that wraps around its surface and came across that error.

lindsayad replied:
Haha look at you coming up with more tricky tests! This is good. If you can, please push up that test, and I will take a look

recuero replied:
Just did!

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1095160/) on 1cf3838 : invalidated by @recuero

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1095169/) on 1cf3838 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -851,15 +857,19 @@ AutomaticMortarGeneration::buildMortarSegmentMesh3d()
       if (secondary_side_elem->subdomain_id() != secondary_subd_id)
         continue;
 
+      auto [secondary_elem_to_msm_map_it, insertion_happened] =
```
This is a structured binding to the returned pair. You have to bind a variable to each member of the returned pair; you can't ignore one of the returned items like you can in python with `_`.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1095169/) on 1cf3838 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)

recuero committed:
Add test with extended sideset.

This test is intended to test the removal of secondary elements from
the mortar segment mesh that do not project. Each secondary element in
the maps must now have corresponding mortar segment mesh elements.

recuero commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -78,8 +80,8 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
-  const auto & its = amg().secondariesToMortarSegments(*_current_node);
-  std::array<MortarNodalAuxKernelTempl<ComputeValueType> *, 1> consumers = {{this}};
+  const std::vector<AutomaticMortarGeneration::MortarFilterIter> & its =
+      amg().secondariesToMortarSegments(*_current_node);
```
I think `auto` tends to make the code less readable, but I am swimming against the tide. Reverted it in the latest commit.

lindsayad committed:
Return empty container if node away from mortar mesh

recuero committed:
Temp commit with wrap around sideset/lower-d

lindsayad committed:
Dont use ad_dofs_dot when it hasn't been computed

lindsayad committed:
Add floating point exception guard for building mortar mesh

lindsayad committed:
If F has no derivatives, then break

lindsayad committed:
Use Lopes appendix code for HH vector computation

recuero committed:
Try adding special case to HH orthogonalization.

If this special case isn't added or the equations do not avoid singularities
for all cases, an FPE will happen.

recuero committed:
Add test for mortar nodal geometry generation

lindsayad committed:
Avoid redundant map find for msm info

lindsayad committed:
Continue instead of breaking when orientations invalid

If you break after one failed projection you may not perform
perfectly valid primary node projections. This led in one case
for me accepting a solution as converged in which an entire block
was penetrated into the primary block

lindsayad committed:
Add back small-volume msm elements

If we remove small-volume msm-elements you can actually have cases
in which a primary node projects into the hole created by not
adding the small-volume element. We can consider removing these
msm elements later if we want, just not where they were removed
here

lindsayad committed:
Revert "Remove unused secondaries from node-to-elem maps"

If we are going to erase from one container, then it seems
like we should erase from all. This incurs additional expense
without a clear benefit to me. It allows more sanity checks
I suppose, but I don't think this is sufficiently beneficial
to outweigh the added cost

This reverts commit 1cf38388c4bb78203a0ded724a1cc265dbcb1ca5.

lindsayad committed:
Add comments regarding nodal-normal/surface orthogonality

lindsayad committed:
Use variable for nodal normals system name

recuero committed:
Add contact test.

It shows that the same contact results can be obtained with a wrap around definition of
 the subdomains but with a sideset that only spans the possible contact area. This is not
true right now for a case where the sideset also wraps around since, when performing projections,
large gaps (not in contact) in many mortar segments have more weight than negative/zero gaps (in contact).
This is a corner case for a square geometry that may need special consideration for the node-based approach
used in mortar mechanical contact.

recuero committed:
Default mortar nodal auxiliary kernel to act on displaced mesh.

recuero committed:
Use only sphere test to check for nodal geometry

lindsayad committed:
Add check for obliqueness of projection

recuero committed:
Add user parameter to reject generation of mortar segments.

- This is useful to avoid extra/meaningless segments.
- Add test to hold behavior for wrap-around lower-d domain
- Wrap-around mesh modifies nodal geometry and therefore numerical results
- Remove nodes if msm elements were removed due to lack of projection

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1105081/) on 39d6275 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21199/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21199/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f15ce9c65307ee843e1d8a58cd0d954f54f65d`

recuero committed:
Fix clang format

Add missing test files.

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -503,6 +505,12 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
   /// Flag to enable regressed treatment of edge dropping where all LM DoFs on edge dropping element
   /// are strongly set to 0.
   bool _correct_edge_dropping;
```
```suggestion
  const bool _correct_edge_dropping;
```

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -503,6 +505,12 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+  /// If set to a value close to zero, very oblique projections are allowed, which can result in mortar
+  /// segments solving physics not meaningfully and overprojection of primary nodes onto the mortar
+  /// segment mesh in extreme cases. This parameter is mostly intended for mortar mesh debugging purposes in 2D.
+  Real _minimum_projection_angle;
```
```suggestion
  const Real _minimum_projection_angle;
```

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -706,6 +707,18 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
     }
   }
 
+  std::unordered_set<Node *> msm_connected_nodes;
+
+  // Deleting elements may produce isolated nodes.
+  // Loops for identifying and removing such nodes from mortar segment mesh.
+  for (const auto & element : _mortar_segment_mesh->element_ptr_range())
+    for (auto & n : element->node_ref_range())
+      msm_connected_nodes.insert(&n);
+
+  for (const auto & node : _mortar_segment_mesh->node_ptr_range())
+    if (!msm_connected_nodes.count(node))
+      _mortar_segment_mesh->delete_node(node);
```
Do these orphaned nodes have a practical impact on the computation? Or does this make a debug mesh or nodal geometry look weird if they're left in? Just curious if there's a motivation for this other than making the mesh "logically clean"

recuero replied:
I don't think they have an impact on the computation; it doesn't for the case I tried out. But those nodes still reside in the `mortar_segment_mesh`, which makes the exodus file (object) different from the case in which we do not wrap the sidesets/lowerds around the block. My main motivation was to ensure the `mortar_segment_mesh` is as close as possible to the reference situation with lower-d only on the contact patch, otherwise, for the bouncing block, e.g., the mesh has double the number of nodes compared to the reference case.

lindsayad replied:
Alright we can keep this then. If it shows up in profiling ever, we can consider removing it

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1572,17 +1591,36 @@ AutomaticMortarGeneration::computeNodalGeometry()
-      std::max(unit_normal(0) - 1.0, unit_normal(0) + 1.0), unit_normal(1), unit_normal(2));
+  const Point h_vector(std::max(nx - 1.0, nx + 1.0), ny, nz);
+
+  if (std::abs(h_vector(0)) < TOLERANCE)
+  {
+    nodal_tangent_one(0) = 0;
+    nodal_tangent_one(1) = 1;
+    nodal_tangent_one(2) = 0;
+
+    nodal_tangent_two(0) = 0;
+    nodal_tangent_two(1) = 0;
+    nodal_tangent_two(2) = -1;
+
+    return;
+  }
```
I assume there was a methodology to this choice? If it's not too complicated, you might consider adding a comment regarding it

recuero replied:
Yeah, the equations yield these vectors as we approach the singularity. Will add a comment to that respect.

recuero replied:
This is an example of how the fields look like on a spherical surface for a node that doesn't go in the avoid-singularity if statement:

<img width="1857" alt="image" src="https://user-images.githubusercontent.com/14300313/178068523-2ef091d5-4e4f-46df-b546-351ecc7cd0ea.png">

lindsayad reviewed (changes requested):
Getting to the finish line I think!

lindsayad commented:
I think it might be worth adding a comment, maybe in the place where we finally delete the msm from the mesh (in 2D), that we may be creating nodes with only one element neighbor where before there were two, but that the nodal normal used in computations will reflect the two-neighbor geometry

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -696,7 +691,13 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+      auto & msm_set = it->second;
+      msm_set.erase(msm_elem);
+      if (msm_set.empty())
+        _secondary_elems_to_mortar_segments.erase(it);
```
Actually I think this would be the most appropriate for the nodal normal comment, since it is the nodes with secondary element neighbors (not mortar segment mesh neighbors) that determine the nodal normal field

recuero committed:
Address Alex's comments

lindsayad reviewed (dismissed):
Maybe we should have @dschwen or someone else do a final run-through since I have a lot of commits in this PR?

dschwen commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -46,7 +48,9 @@ MortarNodalAuxKernelTempl<ComputeValueType>::MortarNodalAuxKernelTempl(
         *this->template getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
     MortarConsumerInterface(this),
-    _displaced(this->template getParam<bool>("use_displaced_mesh")),
+    _displaced(parameters.isParamSetByUser("use_displaced_mesh")
+                   ? this->template getParam<bool>("use_displaced_mesh")
+                   : true),
```
do 

```
params.set<bool>("use_displaced_mesh") = true; 
```
in validParams instead.

dschwen commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -104,6 +109,12 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
+  // to be zero. This can be useful when nodes initially do not project but will project at a
+  // different stage of the simulation
+  if (total_volume == 0)
+    total_volume = std::numeric_limits<Real>::max();
```
why not rather do 

```
  if (total_volume == 0)
    value = 0;
  else
    value /= total_volume;
```
dividing by max real later is a bit of a detour IMO

dschwen replied:
maybe a fuzzy equal?

recuero replied:
We are setting value/total_volume below

dschwen commented on modules/contact/test/tests/3d-mortar-contact/tests:
```diff
@@ -59,19 +59,6 @@
-    requirement = 'The system shall solve a 3D frictionless bouncing block problem with mortar '
-                  'constraint and output the normal and tangent vector fields for debugging purposes'
-    max_parallel = 1
-  []
```
why is this removed?

dschwen commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1572,17 +1597,39 @@ AutomaticMortarGeneration::computeNodalGeometry()
   // Mar 1;45(3):683-94.
-  const Point h_vector(
-      std::max(unit_normal(0) - 1.0, unit_normal(0) + 1.0), unit_normal(1), unit_normal(2));
+  const Point h_vector(std::max(nx - 1.0, nx + 1.0), ny, nz);
```
How is `std::max(nx - 1.0, nx + 1.0)` ever not `nx + 1.0` ?

dschwen commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1727,21 +1775,33 @@ AutomaticMortarGeneration::projectSecondaryNodesSinglePair(
+            else
+              // It's possible that the nodal normal is completely orthogonal to the primary
+              // surface, in which case the derivative is 0. We know in this case that the
+              // projection should be a failure
```
I don't understand this comment. Wouldn't that be a fairly common case in parallel plate contact?

recuero replied:
What's the implication that this case can happen commonly? I don't fully understand your comment.

lindsayad replied:
@dschwen I changed "orthogonal to the primary surface" to "orthogonal to the primary surface normal". Hopefully that clarifies things?

lindsayad replied:
and for good measure "nodal normal" to "secondary surface nodal normal"

dschwen reviewed (commented):
See comments

recuero commented on modules/contact/test/tests/mortar_aux_kernels/block-dynamics-aux-fretting-wear-test-action.i:
```diff
@@ -0,0 +1,283 @@
+[]
+
+[Kernels]
+  [DynamicTensorMechanics]
```
It works. I'll be using it in an upcoming commit, but needed to regold because found differences on the order of 1.0e-3 which can be attributed to the execution of the auxiliary kernels for velocity and acceleration.

recuero committed:
Address Daniel's review

dschwen commented:
`The requested mortar interface AutomaticMortarGeneration object does not yet exist!` something broke. :-(

recuero committed:
Add comments on why not default mortar object to true.

The AugmentSparsityOnInterface object relies on a correct object
setting for _use_displace_mesh. If we override this parameter to
true in cases where there is no displaced mesh, we will be trying to
retrieve a mortar mesh that does not exist. This happens because
FeProblemBase changes the value of this parameter externally.

lindsayad committed:
Clarify orthogonality comment

recuero commented:
I cannot find a way to account for the existence of `displaced_mesh` in the relationship manager parameter of `MortarConsumerInterface`. I believe we'd need that to make the call to `getMortarInterface` from `AugmentSparsityOnInterface` have the correct value for `_use_displaced_mesh`, as this parameter is changed in `FEProblemBase` when creating kernels and auxiliary kernels. Maybe @lindsayad has a better idea.

lindsayad commented:
Why did this issue crop up? (I believe there is an issue, but how are we seeing it now and not before?)

lindsayad commented:
Never mind, I see the most recent commit explains it. That is unfortunate but I don't have a good/easy solution for that 😦

recuero commented:
It's arising now because we are trying to set defaults in use_displaced_mesh for mortar objects that can act on displaced and undisplaced meshes.

lindsayad commented:
it would be nice to get this merged. @dschwen could you check and see if the conversations are resolved?

lindsayad reviewed (approved):
I believe all review comments have been addressed and there is mortar work that may depend on this
